docs/design_proposals/generalizing_kaniko_config.md

```diff
@@ -0,0 +1,137 @@
+# Generalizing Container Registry config for Cluster builds
+
+* Author(s): @venkatk-25
+* Design Shepherd: @balopat
+* Date: 
+* Status: 
+
+## Background
+
+Kaniko supports 2 types of registries GCR and Docker, with former being mandatory always. 
+The implementation itself is repository specific and not extensible as we need to keep adding more logic for every new type of repository and keep maintaining compatibilty with registries/credhelpers.
+
+The current looks like below( I have kept only required pieces from original code snippet):
+``` go
+type ClusterDetails struct {
+       // GCR creds
+	PullSecret     string        `yaml:"pullSecret,omitempty"`
+	PullSecretName string        `yaml:"pullSecretName,omitempty"`
+
+	DockerConfig   *DockerConfig `yaml:"dockerConfig,omitempty"`
+}
+
+// DockerConfig contains information about the docker `config.json` to mount.
+type DockerConfig struct {
+	Path       string `yaml:"path,omitempty"`
+	SecretName string `yaml:"secretName,omitempty"`
+}
+```
+### Breaking down the requirements and generalizing:
+
+If you observe the above piece of configuration, you will see both creds have same 2 fields, SecretName/SecretPath. 
+When we look into how they behave differently, we can observe the differences between them as follows:
+1. The path where to mount the config varies based on secret
+1. Some of them(like GCR) might require env variables to be set.
+
+If we can generalize the above points, then any configuration(GCR/ER/private registry/S3/GCS, etc..) breaks down to following steps:
+1. Create a secret with the configuration
+2. Mount the config in specific location in pod
+3. In some cases set some env variables, for the same.
+
+
+## Design
+
+The same above can be re-configured in a more generic way as follows:
+
+``` go
+type ClusterDetails struct {
+	SecretConfigs []*SecretConfig   `yaml:"secrets,omitempty"`
+	env           map[string]string `yaml:"env,omitempty"`
+}
+
+// DockerConfig contains information about the docker `config.json` to mount.
+type SecretConfig struct {
+	LocalPath  string `yaml:"localPath"`
+	SecretName string `yaml:"secretName,omitempty"`
+	MountPath  string `yaml:"mountPath,omitempty"`
+}
+```
+
+another option for env is to have( thanks @azaiter):
+env: [v1.EnvVar](https://github.com/GoogleCloudPlatform/freshpod/blob/master/vendor/k8s.io/api/core/v1/types.go#L1693)
+### How different configurations fit into this scheme:
+#### GCR/GCS
+current scheme:
+``` yaml
+build:
+  cluster:
+    pullSecretName: e2esecret
+```
+new scheme:
+``` yaml
+build:
+  cluster:
+    secrets: 
+    - secretName: e2esecret
+      mountPath:  "/secret"
+    env:
+       "GOOGLE_APPLICATION_CREDENTIALS" : "/secret/kaniko-secret"
```

Comment:
this still feels like we could generate it - just because the user can make an error easily here - for example this example itself has inconsistency, the env should be: ` "GOOGLE_APPLICATION_CREDENTIALS" : "/secret/e2esecret"`

Reply:
I wasn't sure of this when I wrote it. From code this looked like a constant regardless of secret name. https://github.com/GoogleContainerTools/skaffold/blob/533378cad3a92c391dca9d1656c2c30967c2c5cb/pkg/skaffold/build/kaniko/sources/sources.go#L69-L70

Reply:
> * I would still think about generating/helping the user with the env var somehow for GCR.

Regardless of above comment, I agree this is an additional and error prone step for GCR or maybe other secrets. I can think of following way to solve this:

. Have one more optional field called type for secret which can be an enum of special secrets, so we can do some extra work on them when specified.